packages/rum-core/src/domain/rumEventsCollection/action/rageClickChain.ts

```diff
@@ -0,0 +1,115 @@
+import { monitor, ONE_SECOND, timeStampNow } from '@datadog/browser-core'
+import { FrustrationType } from '../../../rawRumEvent.types'
+import type { Click } from './trackClickActions'
+
+export interface RageClickChain {
+  tryAppend: (click: Click) => boolean
+  stop: () => void
+}
+
+export const MAX_DURATION_BETWEEN_CLICKS = ONE_SECOND
+export const MAX_DISTANCE_BETWEEN_CLICKS = 100
+
+const enum RageClickChainStatus {
+  WaitingForMoreClicks,
+  WaitingForClicksToStop,
+  Finalized,
+}
+
+export function createRageClickChain(firstClick: Click): RageClickChain {
+  const bufferedClicks: Click[] = []
+  let status = RageClickChainStatus.WaitingForMoreClicks
+  let maxDurationBetweenClicksTimeout: number | undefined
+  const rageClick = firstClick.clone()
+
+  function dontAcceptMoreClick() {
+    clearTimeout(maxDurationBetweenClicksTimeout)
+    if (status === RageClickChainStatus.WaitingForMoreClicks) {
+      status = RageClickChainStatus.WaitingForClicksToStop
+      tryFinalize()
+    }
+  }
+
+  function tryFinalize() {
+    if (status === RageClickChainStatus.WaitingForClicksToStop && bufferedClicks.every((click) => click.isStopped())) {
+      status = RageClickChainStatus.Finalized
+      finalizeClicks(bufferedClicks, rageClick)
+    }
+  }
+
+  function appendClick(click: Click) {
+    click.onStop(tryFinalize)
+    bufferedClicks.push(click)
+    clearTimeout(maxDurationBetweenClicksTimeout)
+    maxDurationBetweenClicksTimeout = setTimeout(monitor(dontAcceptMoreClick), MAX_DURATION_BETWEEN_CLICKS)
+  }
+
+  appendClick(firstClick)
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏a bit weird to me that we have those functions declarations between other statements, also discovery order could be more aligned with the workflow, what about:

```suggestion
  const rageClick = firstClick.clone()
  appendClick(firstClick)

  function appendClick(click: Click) {
    click.onStop(tryFinalize)
    bufferedClicks.push(click)
    clearTimeout(maxDurationBetweenClicksTimeout)
    maxDurationBetweenClicksTimeout = setTimeout(monitor(dontAcceptMoreClick), MAX_DURATION_BETWEEN_CLICKS)
  }
  
  function tryFinalize() {
    if (status === RageClickChainStatus.WaitingForClicksToStop && bufferedClicks.every((click) => click.isStopped())) {
      status = RageClickChainStatus.Finalized
      finalizeClicks(bufferedClicks, rageClick)
    }
  }
    
  function dontAcceptMoreClick() {
    clearTimeout(maxDurationBetweenClicksTimeout)
    if (status === RageClickChainStatus.WaitingForMoreClicks) {
      status = RageClickChainStatus.WaitingForClicksToStop
      tryFinalize()
    }
  }
```